File: src/api/src/modules/embedding_generator/validator.py
"""Embedding dimension validation to prevent sync crashes."""

from .types import TEXT_EMBEDDING_DIM, TEXT_MODEL_NAME


class DimensionMismatchError(Exception):
    """Raised when embedding dimensions don't match between systems."""


def validate_embedding_dimension(embedding, context: str = ""):
    """Validate that an embedding has the correct dimension.

    Args:
        embedding: NumPy array or list to validate
        context: Additional context for error message

    Raises:
        DimensionMismatchError: If dimension doesn't match TEXT_EMBEDDING_DIM
    """
    try:
        actual_dim = len(embedding) if hasattr(embedding, "__len__") else embedding.shape[0]
    except (AttributeError, IndexError):
        raise DimensionMismatchError(f"Invalid embedding format: {type(embedding)}")

    if actual_dim != TEXT_EMBEDDING_DIM:
        error_msg = (
            f"Embedding dimension mismatch{': ' + context if context else ''}. "
            f"Expected {TEXT_EMBEDDING_DIM} dimensions (model: {TEXT_MODEL_NAME}), "
            f"but got {actual_dim} dimensions. "
            f"This indicates a configuration mismatch between desktop and backend."
        )
        raise DimensionMismatchError(error_msg)


def validate_embedding_compatibility(desktop_dim: int, backend_dim: int):
    """Validate that desktop and backend embedding dimensions match.

    Args:
        desktop_dim: Embedding dimension from desktop system
        backend_dim: Embedding dimension from backend system

    Raises:
        DimensionMismatchError: If dimensions don't match
    """
    if desktop_dim != backend_dim:
        raise DimensionMismatchError(
            f"CRITICAL: Embedding dimension mismatch detected!\n"
            f"Desktop uses {desktop_dim}-dimensional embeddings.\n"
            f"Backend uses {backend_dim}-dimensional embeddings.\n"
            f"Sync operations will crash until this is fixed.\n"
            f"Both systems must use the same embedding model and dimension.\n"
            f"Current expected: {TEXT_EMBEDDING_DIM} dimensions with model {TEXT_MODEL_NAME}"
        )


def get_expected_dimension() -> int:
    """Get the expected embedding dimension for this system."""
    return TEXT_EMBEDDING_DIM


def get_expected_model() -> str:
    """Get the expected embedding model name for this system."""
    return TEXT_MODEL_NAME

File: src/app/src/src/features/embedding/validator.rs
use super::generator::ModelConfig;
use crate::shared::error::{AppError, Result};

/// Error raised when embedding dimensions don't match between systems
#[derive(Debug, thiserror::Error)]
#[error("Embedding dimension mismatch: {message}")]
pub struct DimensionMismatchError {
    pub message: String,
}

impl From<DimensionMismatchError> for AppError {
    fn from(err: DimensionMismatchError) -> Self {
        AppError::EmbeddingFailed {
            reason: err.message,
        }
    }
}

/// Validate that an embedding has the correct dimension
///
/// # Arguments
/// * `embedding` - The embedding vector to validate
/// * `expected_dim` - Expected dimension
/// * `context` - Additional context for error message
///
/// # Errors
/// Returns error if dimension doesn't match
pub fn validate_embedding_dimension(
    embedding: &[f32],
    expected_dim: usize,
    context: &str,
) -> Result<()> {
    let actual_dim = embedding.len();

    if actual_dim != expected_dim {
        let error_msg = if context.is_empty() {
            format!(
                "Embedding dimension mismatch. Expected {} dimensions, but got {} dimensions.",
                expected_dim, actual_dim
            )
        } else {
            format!(
                "Embedding dimension mismatch: {}. Expected {} dimensions, but got {} dimensions.",
                context, expected_dim, actual_dim
            )
        };

        return Err(DimensionMismatchError { message: error_msg }.into());
    }

    Ok(())
}

/// Validate that desktop and backend embedding dimensions match
///
/// # Arguments
/// * `desktop_dim` - Embedding dimension from desktop system
/// * `backend_dim` - Embedding dimension from backend system
///
/// # Errors
/// Returns error if dimensions don't match
pub fn validate_embedding_compatibility(
    desktop_dim: usize,
    backend_dim: usize,
    model_config: &ModelConfig,
) -> Result<()> {
    if desktop_dim != backend_dim {
        let error_msg = format!(
            "CRITICAL: Embedding dimension mismatch detected!\n\
            Desktop uses {}-dimensional embeddings.\n\
            Backend uses {}-dimensional embeddings.\n\
            Sync operations will crash until this is fixed.\n\
            Both systems must use the same embedding model and dimension.\n\
            Current expected: {} dimensions with model {}",
            desktop_dim,
            backend_dim,
            model_config.dimensions(),
            model_config.model_name()
        );

        return Err(DimensionMismatchError { message: error_msg }.into());
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::embedding_constants::DEFAULT_EMBEDDING_DIM;

    #[test]
    fn test_validate_embedding_dimension_success() {
        let embedding = vec![0.1, 0.2, 0.3];
        let result = validate_embedding_dimension(&embedding, 3, "test");
        assert!(result.is_ok());
    }

    #[test]
    fn test_validate_embedding_dimension_failure() {
        let embedding = vec![0.1, 0.2, 0.3];
        let result = validate_embedding_dimension(&embedding, 5, "test context");
        assert!(result.is_err());

        let err_msg = result.unwrap_err().to_string();
        assert!(err_msg.contains("test context"));
        assert!(err_msg.contains("Expected 5"));
        assert!(err_msg.contains("got 3"));
    }

    #[test]
    fn test_validate_compatibility_success() {
        let config = ModelConfig::AllMpnetBaseV2;
        let result =
            validate_embedding_compatibility(DEFAULT_EMBEDDING_DIM, DEFAULT_EMBEDDING_DIM, &config);
        assert!(result.is_ok());
    }

    #[test]
    fn test_validate_compatibility_failure() {
        let config = ModelConfig::AllMpnetBaseV2;
        let result = validate_embedding_compatibility(DEFAULT_EMBEDDING_DIM, 384, &config);
        assert!(result.is_err());

        let err_msg = result.unwrap_err().to_string();
        assert!(err_msg.contains("CRITICAL"));
        assert!(err_msg.contains("384"));
        assert!(err_msg.contains(&DEFAULT_EMBEDDING_DIM.to_string()));
    }
}
